src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonParameterInfo.cs

```diff
@@ -11,108 +11,43 @@ namespace System.Text.Json.Serialization.Metadata
     /// </summary>
     internal abstract class JsonParameterInfo
     {
-        private JsonTypeInfo? _jsonTypeInfo;
+        public JsonConverter EffectiveConverter => MatchingProperty.EffectiveConverter;
 
-        public JsonConverter ConverterBase { get; private set; } = null!;
+        // The default value of the parameter. This is `DefaultValue` of the `ParameterInfo`, if specified, or the `default` for the `ParameterType`.
+        public object? DefaultValue { get; private protected init; }
 
-        private protected bool MatchingPropertyCanBeNull { get; private set; }
+        public bool IgnoreNullTokensOnRead { get; }
 
-        // The default value of the parameter. This is `DefaultValue` of the `ParameterInfo`, if specified, or the CLR `default` for the `ParameterType`.
-        public object? DefaultValue { get; private protected set; }
-
-        public bool IgnoreNullTokensOnRead { get; private set; }
-
-        // Options can be referenced here since all JsonPropertyInfos originate from a JsonTypeInfo that is cached on JsonSerializerOptions.
-        public JsonSerializerOptions? Options { get; set; } // initialized in Init method
+        public JsonSerializerOptions Options { get; }
 
         // The name of the parameter as UTF-8 bytes.
-        public byte[] NameAsUtf8Bytes { get; private set; } = null!;
+        public byte[] NameAsUtf8Bytes { get; }
 
-        public JsonNumberHandling? NumberHandling { get; private set; }
+        public JsonNumberHandling? NumberHandling { get; }
 
-        // Using a field to avoid copy semantics.
-        public JsonParameterInfoValues ClrInfo = null!;
+        public int Position { get; }
 
-        public JsonTypeInfo JsonTypeInfo
-        {
-            get
-            {
-                Debug.Assert(Options != null);
-                Debug.Assert(ShouldDeserialize);
-                return _jsonTypeInfo ??= Options.GetTypeInfoInternal(PropertyType);
-            }
-            set
-            {
-                // Used by JsonMetadataServices.
-                Debug.Assert(_jsonTypeInfo == null);
-                _jsonTypeInfo = value;
-            }
-        }
+        public JsonTypeInfo JsonTypeInfo => MatchingProperty.JsonTypeInfo;
 
-        public Type PropertyType { get; set; } = null!;
+        public Type ParameterType { get; }
 
-        public bool ShouldDeserialize { get; private set; }
+        public bool ShouldDeserialize { get; }
 
-        public JsonPropertyInfo MatchingProperty { get; private set; } = null!;
+        public JsonPropertyInfo MatchingProperty { get; }
 
-        public virtual void Initialize(JsonParameterInfoValues parameterInfo, JsonPropertyInfo matchingProperty, JsonSerializerOptions options)
```

Comment:
Replaces `Initialize` methods with constructors -- makes it easier to guarantee non-nullability of properties.